File: step34_event-driven-restaurant-app/Python/cdk-backend/lambda-fns/dynamoHandler.py
# import { randomBytes } from 'crypto';
import boto3
import os
import secrets
import json
from botocore.exceptions import ClientError
dynamodb = boto3.resource('dynamodb')
table = os.environ['DYNAMO_TABLE_NAME']
dbtable= dynamodb.Table(table)

def handler(event, context):

    print(event)
    returningPayload = json.dumps({ "operationSuccessful" : True })
    try:
        # ///  adding new time slot ////
        if event["detail-type"] == "addTimeSlot":
            res = dbtable.put_item(
                Item= { "id" : secrets.token_hex(16), **event['detail'], "isBooked" : False },
               )
            print(json.dumps(res))

        # ///  deleting time slot ///// 
        if event["detail-type"] == "deleteTimeSlot":
            res = dbtable.delete_item(
               Key={ 'id': event['detail']['id'] }
            )
            print(json.dumps(res)) 

        #  ////  adding booking time slot request ////
        if event["detail-type"] == "addBookingRequest":
            res = dbtable.update_item(
                  Key={ "id": event['detail']['id'] },
                  UpdateExpression= "set isBookingRequested = :booleanValue, bookingRequestBy = :userName",
                 ExpressionAttributeValues= {
                    ":booleanValue": True,
                    ":userName": event['detail']['userName']
                },
                ReturnValues= "UPDATED_NEW" 
            )
            returningPayload= json.dumps({ "SnsMessage" : 'Request of booking timeSlot', "operationSuccessful" : True  })
  
            print(json.dumps(res)) 

        # ////  deleting booking time slot request ///  
        if event["detail-type"] == "deleteBookingRequest":
            res = dbtable.update_item(
                Key= { "id": event['detail']['id'] },
                UpdateExpression= "set isBookingRequested = :booleanValue, bookingRequestBy = :userName",
                ExpressionAttributeValues= {
                    ":booleanValue": False,
                    ":userName": ''
                },
                ReturnValues= "UPDATED_NEW"  )
            returningPayload= json.dumps({ "SnsMessage" : 'Request of booking timeSlot', "operationSuccessful" : True  })

            print(json.dumps(res)) 

        #/////  canceling booked time slot ///// 
        if event["detail-type"] == "cancelBooking":
            res = dbtable.update_item(

                Key= { "id": event['detail']['id'] },
                UpdateExpression= "set isBooked = :_isBooked, bookedBy = :_bookedBy",
                ExpressionAttributeValues= {
                    ":_isBooked": False,
                    ":_bookedBy": "",
                },
                ReturnValues= "UPDATED_NEW"   )
            returningPayload= json.dumps({ "SnsMessage" : 'Request of canceling timeSlot', "operationSuccessful" : True  })    

            print(json.dumps(res)) 

        # ////  canceling all booked time slots ////
        if event["detail-type"] == "resetAllBookings":
            res = dbtable.scan()
            data = res['Items']
            print(data)
            for d in data:
                # print("ID ===>> ", d["id"])
                output = dbtable.update_item(
                    Key= { "id": d["id"] },
                    UpdateExpression= "set isBooked = :_isBooked, bookedBy = :_bookedBy, isBookingRequested = :_isBookingRequested, bookingRequestBy = :_bookingRequestBy",
                    ExpressionAttributeValues={
                        ":_isBooked": False,
                        ":_bookedBy": "",
                        ":_isBookingRequested": False,
                        ":_bookingRequestBy": ""
                    },
                    ReturnValues= "NONE" 
                )
                print(json.dumps(output))  

        # /////  booking time slot ///////
        if event["detail-type"] == "bookTimeSlot":
            data1 = dbtable.get_item(
                Key= { 'id': event['detail']['id'] },
                AttributesToGet= ["isBookingRequested", "bookingRequestBy"]
            )
            print(data1)
            if data1['Item']['isBookingRequested'] == True:
                res = dbtable.update_item(
                    Key= { "id": event['detail']['id'] },
                    UpdateExpression="set isBooked = :_isBooked, bookedBy = :_bookedBy, isBookingRequested = :_isBookingRequested, bookingRequestBy = :_bookingRequestBy",
                     ExpressionAttributeValues= {
                        ":_isBooked": True,
                        ":_bookedBy": data['Item']['bookingRequestBy'], 
                        ":_isBookingRequested": False,
                        ":_bookingRequestBy": ""
                    },
                     ReturnValues= "UPDATED_NEW" 
                )
                print(res)
                print(json.dumps(res))
                returningPayload= json.dumps({ "SnsMessage" : 'Request of booking timeSlot', "operationSuccessful" : True  }) 
        
              
        print("returning payload", returningPayload)
        return returningPayload;
       
    except ClientError as  e:
        print(e)
        returningPayload['operationSuccessful'] = False
        return returningPayload
        
    

File: step34_event-driven-restaurant-app/cdk-backend/lambda-fns/dynamoHandler.ts
import { EventBridgeEvent, Context } from 'aws-lambda';
import { randomBytes } from 'crypto';
import * as AWS from 'aws-sdk';

const dynamoClient = new AWS.DynamoDB.DocumentClient();
const TABLE_NAME = process.env.DYNAMO_TABLE_NAME as string;

export type PayloadType = {
    operationSuccessful: boolean,
    SnsMessage?: string,
    customerEmail?:string
}

export const handler = async (event: EventBridgeEvent<string, any>, context: Context) => {

    console.log(JSON.stringify(event, null, 2));
    const returningPayload: PayloadType = { operationSuccessful: true };

    try {
        //////////////  adding new time slot /////////////////////////
        if (event["detail-type"] === "addTimeSlot") {
            // console.log("detail===>", JSON.stringify(event.detail, null, 2));
            const params = {
                TableName: TABLE_NAME,
                Item: { id: randomBytes(16).toString("hex"), ...event.detail, isBooked: false },
            }
            await dynamoClient.put(params).promise();
        }

     
        //////////////  booking time slot /////////////////////////
        else if (event["detail-type"] === "bookTimeSlot") {
            // geting the time slot by id
            const data = await dynamoClient.get({
                TableName: TABLE_NAME,
                Key: { id: event.detail.id },
                AttributesToGet: ["isBookingRequested", "bookingRequestBy"]
            }).promise();

            // if the time slot has a booking request than booked for that user
            if (data.Item?.isBookingRequested) {
                const params = {
                    TableName: TABLE_NAME,
                    Key: { "id": event.detail.id },
                    UpdateExpression: "set isBooked = :_isBooked, bookedBy = :_bookedBy, isBookingRequested = :_isBookingRequested, bookingRequestBy = :_bookingRequestBy",
                    ExpressionAttributeValues: {
                        ":_isBooked": true,
                        ":_bookedBy": data.Item?.bookingRequestBy, // userName
                        ":_isBookingRequested": false,
                        ":_bookingRequestBy": ""
                    },
                    ReturnValues: "UPDATED_NEW" // NONE | ALL_OLD | UPDATED_OLD | ALL_NEW | UPDATED_NEW,
                };
              const result:any =  await dynamoClient.update(params).promise();
         
              returningPayload.SnsMessage = "your booking request has been accepted"
                returningPayload.customerEmail = result.Attributes.bookingRequestBy;
                console.log(event["detail-type"])
                console.log(returningPayload);

                return returningPayload

            }
        }

        //////////////  adding booking time slot request /////////////////////////
        else if (event["detail-type"] === "addBookingRequest") {
            const params = {
                TableName: TABLE_NAME,
                Key: { "id": event.detail.id },
                UpdateExpression: "set isBookingRequested = :booleanValue, bookingRequestBy = :userName",
                ExpressionAttributeValues: {
                    ":booleanValue": true,
                    ":userName": event.detail.userName
                },
                ReturnValues: "UPDATED_NEW" // NONE | ALL_OLD | UPDATED_OLD | ALL_NEW | UPDATED_NEW,
            };
            await dynamoClient.update(params).promise();

            // adding sns message
            returningPayload.SnsMessage = 'Request of booking timeSlot';

            console.log(event["detail-type"])
            console.log(returningPayload);
        }

        //////////////  deleting booking time slot request /////////////////////////
        else if (event["detail-type"] === "deleteBookingRequest") {
            const params = {
                TableName: TABLE_NAME,
                Key: { "id": event.detail.id },
                UpdateExpression: "set isBookingRequested = :booleanValue, bookingRequestBy = :userName",
                ExpressionAttributeValues: {
                    ":booleanValue": false,
                    ":userName": ''
                },
                ReturnValues: "UPDATED_OLD" // NONE | ALL_OLD | UPDATED_OLD | ALL_NEW | UPDATED_NEW,
            };
          const result:any=   await dynamoClient.update(params).promise();
          

          returningPayload.SnsMessage = "your booking request has been declined"
         returningPayload.customerEmail = result.Attributes.bookingRequestBy;

         console.log(returningPayload);
         console.log(event["detail-type"])
         return returningPayload
     

        }

        console.log(returningPayload);

        //////////////////////////////////  Returning Final Response ///////////////////////////////////////
        return returningPayload;

    } catch (error) {
        console.log(returningPayload);

        console.log("ERROR ====>", error);
        returningPayload.operationSuccessful = false;
        return returningPayload;

    }




};